bot/cogs/reminders.py

```diff
@@ -152,11 +140,10 @@ def ensure_valid_reminder(
                 name=f"Sorry it arrived {humanize_delta(late, max_units=2)} late!"
             )
 
-        await channel.send(
-            content=user.mention,
-            embed=embed
-        )
-        await self._delete_reminder(reminder["id"])
+        await channel.send(content=user.mention, embed=embed)
+
+        log.debug(f"Deleting reminder #{reminder['id']} (the user has been reminded).")
+        await self.bot.api_client.delete(f"bot/reminders/{reminder['id']}")
```

Comment:
Because edited reminders re-use the previous ID, this delete could lead to a race condition:

1. Original reminder starts being sent
2. User edits reminder and PATCH request goes through
3. Original reminder is sent and DELETE request goes through
4. Original reminder task is cancelled (well, too late - it's already finished and been deleted) and edited one is scheduled

Now the reminder is scheduled but doesn't actually exist in the database. I wanted to put this delete into a `finally` block to ensure it's deleted even if it fails to send on Discord, but that would lead to even worse problems. In that case, the reminder would still be deleted even if the task gets cancelled.

A fix may be to first cancel the task, then send the PATCH request. However, I'm don't know how to be certain the task has been cancelled besides accessing the internal scheduler dict and polling for the task to be cancelled. The edit commands somehow need to wait for the task to be cancelled before proceeding. I'm not sure if it's even import to be so certain it's cancelled. It causing an issue may be a really small chance, though pretty sure it'd be more relevant if the delete was in the `finally` block like I mentioned earlier.

Another fix would be to create a new ID for the reminder. Then, the original reminder being deleted won't affect the edited reminder. This seems like a simple and robust solution but perhaps at the cost of convenience for users.

Ideas for solutions are welcome.

Reply:
I've solved this by creating a decorator to abort other commands which try to use the same reminder.